File: java-version/src/main/java/bernard/tatin/threads/AThConsumer.java
package bernard.tatin.threads;


import bernard.tatin.common.IThConsumer;

public abstract class AThConsumer implements IThConsumer, Runnable {
    public static final ProtectedValue<Boolean> isRunning = new ProtectedValue<Boolean>(true);

    @Override
    public void initialize() {
        Thread theThread = new Thread(this, getName());

        theThread.setDaemon(true);
        theThread.start();
    }

    @Override
    public synchronized void run() {
        consume();
    }

    @Override
    public void consume() {
        while (AThConsumer.isRunning.get()) {
            innerLoop();
        }
    }
}

File: kotlin-version/src/main/kotlin/bernard/tatin/Threads/AThConsumer.kt
package bernard.tatin.Threads

abstract class AThConsumer : IThConsumer, Runnable {

    @Override
    fun initialize() {
        val theThread = Thread(this, "AThConsumer")

        theThread.setDaemon(true)
        theThread.start()
    }

    @Override
    @Synchronized
    fun run() {
        consume()
    }

    @Override
    fun consume() {
        while (AThConsumer.isRunning.get()) {
            innerLoop()
        }
    }

    companion object {
        val isRunning = ProtectedFlag(true)
    }
}
